File: probs/208.cpp
#include <string>
using namespace std;

class Vertex {
public:
    Vertex* children[26];
    bool isWord;
};

class Trie {
public:
    Trie() {
        root = new Vertex();
    }
    
    void insert(string word) {
        Vertex* curr = root;

        for (char c: word) {
            if (curr->children[c - 'a'] == nullptr) {
                curr->children[c - 'a'] = new Vertex();
            }

            curr = curr->children[c - 'a'];
        }

        curr->isWord = true;
    }
    
    bool search(string word) {
        Vertex* curr = root;

        for (char c: word) {
            if (curr->children[c - 'a'] == nullptr) {
                return false;
            }

            curr = curr->children[c - 'a'];
        }

        return curr->isWord;
    }
    
    bool startsWith(string prefix) {
        Vertex* curr = root;

        for (char c: prefix) {
            if (curr->children[c - 'a'] == nullptr) {
                return false;
            }

            curr = curr->children[c - 'a'];
        }

        return true;
    }

private:
    Vertex* root;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: probs/208.java
class Trie {
    class Vertex {
        Vertex[] children;
        boolean isWord;

        private Vertex() {
            this.children = new Vertex[26];
        }
    }

    Vertex root;

    public Trie() {
        this.root = new Vertex();
    }
    
    public void insert(String word) {
        Vertex curr = this.root;

        for (char c: word.toCharArray()) {
            if (curr.children[c - 'a'] == null) {
                curr.children[c - 'a'] = new Vertex();
            }

            curr = curr.children[c - 'a'];
        }

        curr.isWord = true;
    }
    
    public boolean search(String word) {
        Vertex curr = this.root;

        for (char c: word.toCharArray()) {
            if (curr.children[c - 'a'] == null) {
                return false;
            }

            curr = curr.children[c - 'a'];
        }

        return curr.isWord;
    }
    
    public boolean startsWith(String prefix) {
        Vertex curr = this.root;

        for (char c: prefix.toCharArray()) {
            if (curr.children[c - 'a'] == null) {
                return false;
            }

            curr = curr.children[c - 'a'];
        }

        return true;
    }
}
